### PR TITLE
bpo-44498: Issue a deprecation warning on asynchat, asyncore and smtpd import

### PR DESCRIPTION
**Pull request by warsaw**
https://bugs.python.org/issue44498

### CODE REVIEW

**Comment by iritkatriel** [comment]
I think we need to add them for asyncore and asynchat as well

**Comment by warsaw** [comment]
I was trying to be targeted in this PR, but it's easy enough to add.

**Comment by iritkatriel** [comment]
Maybe we should ask @pablogsal whether he's happy to backport all of them to 3.10 (or only some)?

**Comment by warsaw** [comment]
@pablogsal - How do you feel about that?  I added the 3.10 backport label in anticipation of your approval :smile:

**Review by iritkatriel — APPROVED** [review]
LGTM.

**Comment by iritkatriel** [comment]
There are tests that use these modules, is that an issue?

**Comment by pablogsal** [comment]
> @pablogsal - How do you feel about that?  I added the 3.10 backport label in anticipation of your approval :smile:

Approval granted ✨

**Review by pablogsal — APPROVED** [review]

**Comment by pablogsal** [comment]
Are these reflected in the 3.10 what's new?

**Comment by pablogsal** [comment]
> There are tests that use these modules, is that an issue?

It may be when running the test suite with -W error so please also check that before landing

**Comment by iritkatriel** [comment]
> > There are tests that use these modules, is that an issue?
> 
> It may be when running the test suite with -W error so please also check that before landing

I'll check the tests.

**Comment by Mariatta** [comment]
Both [asynchat](https://docs.python.org/3/library/asynchat.html) and [asyncore](https://docs.python.org/3/library/asyncore.html) already says "Deprecated since version 3.6" at the top of their documentation.

However there is not yet such a note on the doc of [smtpdlib](https://docs.python.org/3/library/smtpd.html), so we should add it too.

**Comment by iritkatriel** [comment]
> Both [asynchat](https://docs.python.org/3/library/asynchat.html) and [asyncore](https://docs.python.org/3/library/asyncore.html) already says "Deprecated since version 3.6" at the top of their documentation.
> 
> However there is not yet such a note on the doc of [smtpdlib](https://docs.python.org/3/library/smtpd.html), so we should add it too.

There is a comment saying "smtpd should be considered deprecated." in they yellow box  (it doesn't mention the 3.6 version, but it's in the docs for all versions back to 3.5).

**Comment by Mariatta** [comment]
In my mind, "should be considered deprecated" and "Deprecated since " have different meaning.
Since we're adding the code to properly warn users about this being deprecated, I think we should use the ``deprecated`` directive to make it official in the docs. 

```
.. deprecated:: 3.10
   Please use aiosmtpd instead. 
```

or something like that.

**Comment by warsaw** [comment]
> There are tests that use these modules, is that an issue?

I just switched `test_support.py` to use `pprint` instead of `asyncore`.  I think the rest of them are for modules that are also being deprecated, except `smtplib`.  We'll have to think about that one a bit.

**Comment by warsaw** [comment]
> Are these reflected in the 3.10 what's new?

I just added some text about it.

**Comment by warsaw** [comment]
> In my mind, "should be considered deprecated" and "Deprecated since " have different meaning.
> Since we're adding the code to properly warn users about this being deprecated, I think we should use the `deprecated` directive to make it official in the docs.
> 
> ```
> .. deprecated:: 3.10
>    Please use aiosmtpd instead. 
> ```
> 
> or something like that.

Essentially we had that with a `seealso`.  I changed it to `deprecated`

File: Lib/smtpd.py [hunk 1]
```diff
@@ -86,6 +86,13 @@
     "SMTPChannel", "SMTPServer", "DebuggingServer", "PureProxy",
 ]
 
+warn(
+    'The smtpd module is deprecated and unmaintained.  Please see aiosmtpd '
+    '(http://aiosmtpd.readthedocs.io/) for the recommended replacement.',
```
Comment (hroncok) [inline Lib/smtpd.py]:
```suggestion
    '(https://aiosmtpd.readthedocs.io/) for the recommended replacement.',
```

File: Doc/library/smtpd.rst [hunk 2]
```diff
@@ -13,11 +13,10 @@
 
 This module offers several classes to implement SMTP (email) servers.
 
-.. seealso::
-
-    The `aiosmtpd <http://aiosmtpd.readthedocs.io/>`_ package is a recommended
-    replacement for this module.  It is based on :mod:`asyncio` and provides a
-    more straightforward API.  :mod:`smtpd` should be considered deprecated.
+.. deprecated:: 3.6
+   The `aiosmtpd <http://aiosmtpd.readthedocs.io/>`_ package is a recommended
```
Comment (hroncok) [inline Doc/library/smtpd.rst]:
```suggestion
   The `aiosmtpd <https://aiosmtpd.readthedocs.io/>`_ package is a recommended
```

**Review by hroncok — COMMENTED** [review]
I suggest to use HTTPS URLs where available.

File: Lib/smtpd.py [hunk 3]
```diff
@@ -86,6 +86,13 @@
     "SMTPChannel", "SMTPServer", "DebuggingServer", "PureProxy",
 ]
 
+warn(
+    'The smtpd module is deprecated and unmaintained.  Please see aiosmtpd '
+    '(https://aiosmtpd.readthedocs.io/) for the recommended replacement.',
+    DeprecationWarning,
+    stacklevel=2)
```
Comment (iritkatriel) [inline Lib/smtpd.py:91]: This should probably be before the import asyncore. At the moment if you import smtpd you get a deprecation warning for asyncore, which is confusing (particularly with -Derror, where it exits without proceeding to give the error for smtpd as well).

Comment (warsaw) [reply Lib/smtpd.py:91]: Good point, thanks.

**Pull request merged**